Add support for filtering tests using a regular expression

rhmdnd opened the pull request:
Golang supports filtering tests using a -run flag, which can be a
regular expression. The e2e tests don't honor this flag since we're
building the tests in a particular way so we can run parallel tests
(e.g., non-destructive) tests before running the serial tests, which
test remediations and other actions that affect the state of the
cluster, potentially affecting other tests.

To make things more egonomic for contributors, it would be nice to run a
single test so you don't have to wait for the entire suite to execute.

This commit adds a framework flag that introduces a new command line
argument called `-testRegex` and filters e2e tests using the flag as if
it were a regular expression.

Fixes https://github.com/ComplianceAsCode/compliance-operator/issues/226

rhmdnd commented:
Holding this since I tried reusing `-run` but that's going to conflict with the `t.Run()` implementation.

Vincent056 reviewed (approved):
Just tested with the patch, this works, thanks for adding this.

/lgtm

rhmdnd commented:
Just an update - I'm still researching ways to accomplish this without having to roll our own argument.

If folks are fine with this approach knowing it might change in the future, we can remove the hold.

JAORMX commented:
@rhmdnd can you tell me what invocation of `go test ... -run ...` did you try out? I still think this should work out of the box with our usage of `t.Run`.

jhrozek commented:
needs rebase

rhmdnd commented:
> @rhmdnd can you tell me what invocation of `go test ... -run ...` did you try out? I still think this should work out of the box with our usage of `t.Run`.

It does. This patch isn't needed anymore. I proposed an alternative to update the examples that contributors would use to actually run a single test.

https://github.com/ComplianceAsCode/compliance-operator/pull/236

Based on the example in `Makefile`, I was attempting to use derivatives of `TestE2E/TestScanWithNodeSelectorFiltersCorrectly` or `TestScanWithNodeSelectorFiltersCorrectly`, but neither of those worked. Maybe the `Makefile` example predates the support for splitting tests into parallel and serial tests?

If the test name comes from whatever is in `t.Run()`, then the name makes sense. It just wasn't clear from the examples.

https://github.com/ComplianceAsCode/compliance-operator/blob/master/tests/e2e/helpers.go#L290

rhmdnd commented:
That said - I thought the issue was in how we were nesting `t.Run()` calls, which led me to experiment with an alternative structure that would flatten the tests:

https://github.com/ComplianceAsCode/compliance-operator/pull/234

The primary things I focused on was 1.) separating parallel from serial tests 2.) using testing.M to allow for global setup and teardown (instead of using TestE2E), and 3.) using t.Cleanup() instead of defer for cleaning up resources:

https://github.com/golang/go/issues/32111

I thought 1 would be nice to have separate CI jobs, one for parallel and one for serial, although there isn't anything stopping us from doing this today.

I incorporated 2 because I wanted to flatten the test structure, so I needed to find a replacement for TestE2E, which acts as a setup suite. Based on some research, it seemed like using testing.M was the preferred way of setting up and tearing down global state for a collection of tests.

I incorporated 3 because I noticed defer was getting called before some parallel tests executed in an example. It appears based on discussion in the upstream community, that it's preferred to use `t.Cleanup()` over `defer` for cleaning up test resources.

I'm not sure if any of those are still valuable to pursue.
